File: app/src/main/java/com/github/ykiselev/playground/services/schedule/AppSchedule.java
/*
 * Copyright 2017 Yuriy Kiselev (uze@yandex.ru)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.ykiselev.playground.services.schedule;

import com.github.ykiselev.spi.services.schedule.Repeatable;
import com.github.ykiselev.spi.services.schedule.Schedule;

import java.util.PriorityQueue;
import java.util.concurrent.TimeUnit;
import java.util.function.LongSupplier;

import static java.util.Objects.requireNonNull;

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
public final class AppSchedule implements Schedule, AutoCloseable {

    private final PriorityQueue<Task> tasks = new PriorityQueue<>();

    private final LongSupplier clock;

    public AppSchedule(LongSupplier clock) {
        this.clock = requireNonNull(clock);
    }

    public AppSchedule() {
        this(System::currentTimeMillis);
    }

    @Override
    public void close() {
        // no-op for now
    }

    private void add(Task t) {
        synchronized (tasks) {
            tasks.add(t);
        }
    }

    @Override
    public void schedule(long interval, TimeUnit unit, Runnable task) {
        add(task(unit.toMillis(interval), task));
    }

    @Override
    public void schedule(long interval, TimeUnit unit, Repeatable task) {
        add(task(unit.toMillis(interval), task));
    }

    private long time() {
        return clock.getAsLong();
    }

    private Task task(long interval, Runnable r) {
        return new Task(time() + interval, interval, r);
    }

    private Task task(long interval, Repeatable r) {
        return new Task(time() + interval, interval, r);
    }

    @Override
    public void processPendingTasks(long quota) {
        final long t0 = time();
        for (; ; ) {
            final long time = time();
            final long timeLeft = t0 + quota - time;
            if (timeLeft <= 0) {
                break;
            }
            final Task task;
            synchronized (tasks) {
                task = tasks.peek();
                if (task == null || task.after(time + timeLeft)) {
                    break;
                }
                tasks.poll();
            }
            if (task.run()) {
                add(task.reset(time()));
            }
        }
    }

    private static final class Task implements Comparable<Task>, Repeatable {

        private final long targetTime;

        private final long interval;

        private final Repeatable repeatable;

        Task(long targetTime, long interval, Repeatable repeatable) {
            this.targetTime = targetTime;
            this.interval = interval;
            this.repeatable = requireNonNull(repeatable);
        }

        Task(long targetTime, long interval, Runnable runnable) {
            this(targetTime, interval, () -> {
                runnable.run();
                return false;
            });
        }

        Task reset(long time) {
            return new Task(time + interval, interval, repeatable);
        }

        boolean after(long time) {
            return targetTime > time;
        }

        @Override
        public int compareTo(Task o) {
            return Long.compare(targetTime, o.targetTime);
        }

        @Override
        public boolean run() {
            return repeatable.run();
        }
    }

}

File: app/src/main/kotlin/com/github/ykiselev/playground/services/schedule/AppSchedule.kt
package com.github.ykiselev.playground.services.schedule

import com.github.ykiselev.spi.services.schedule.Repeatable
import com.github.ykiselev.spi.services.schedule.Schedule
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.function.LongSupplier

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
class AppSchedule(
    private val clock: LongSupplier = LongSupplier { System.currentTimeMillis() }
) : Schedule, AutoCloseable {

    private val tasks = PriorityQueue<Task>()

    override fun close() {
        // no-op for now
    }

    private fun add(t: Task) {
        synchronized(tasks) { tasks.add(t) }
    }

    override fun schedule(interval: Long, unit: TimeUnit, task: Runnable) {
        add(task(unit.toMillis(interval), task))
    }

    override fun schedule(interval: Long, unit: TimeUnit, task: Repeatable) {
        add(task(unit.toMillis(interval), task))
    }

    private fun time(): Long =
        clock.asLong

    private fun task(interval: Long, r: Runnable): Task =
        Task(time() + interval, interval, r)

    private fun task(interval: Long, r: Repeatable): Task =
        Task(time() + interval, interval, r)

    override fun processPendingTasks(quota: Long) {
        val t0 = time()
        while (true) {
            val time = time()
            val timeLeft = t0 + quota - time
            if (timeLeft <= 0) {
                break
            }
            val task: Task?
            synchronized(tasks) {
                task = tasks.peek()
                if (task == null || task.after(time + timeLeft)) {
                    return
                }
                tasks.poll()
            }
            if (task?.run() == true) {
                add(task.reset(time()))
            }
        }
    }

    private class Task(
        private val targetTime: Long,
        private val interval: Long,
        private val repeatable: Repeatable
    ) : Comparable<Task>, Repeatable {

        constructor(targetTime: Long, interval: Long, runnable: Runnable) : this(
            targetTime,
            interval,
            Repeatable {
                runnable.run()
                false
            })

        fun reset(time: Long): Task =
            Task(time + interval, interval, repeatable)

        fun after(time: Long): Boolean =
            targetTime > time

        override fun compareTo(o: Task): Int =
            targetTime.compareTo(o.targetTime)

        override fun run(): Boolean =
            repeatable.run()
    }
}